no issue - much more consistent vendor name and version api (#22)
makinacorpus/php-query-builder: docs/content/introduction/vendor-conditionals.md

# Database vendor version conditionals

## Introduction

When you write portable code, you might want to dynamically branch your code
depending upon the database server vendor name or version. This library exposes
a few helpers method for doing that gracefully.

:::info
Version string must always be a semantic version string, in the `x.y.z`
(major, minor, patch) form where `x`, `y` and `z` are integers.
You can reduce it to either `x.y` or simply `x` whenever it suits you,
missing digits will always be replaced by `0`.
:::

This section documents methods on the `MakinaCorpus\QueryBuilder\DatabaseSession`

> interface. All briges implement it.

Reviewer: ```suggestion
interface. All bridges implement it.
```

## Compare database vendor name

Simply use the `DatabaseSession::vendorIs()` method, such as:

```php
use MakinaCorpus\QueryBuilder\DatabaseSession;
use MakinaCorpus\QueryBuilder\Platform;

\assert($session instanceof DatabaseSession);

> if ($session->vendorIs(Platform::MYSQL)) {

Author: Vendor name is the name given by the vendor, for example the real vendor for MySQL is Oracle (the company name). But the protocol and tool is the platform. So I'm still not convinced the `Platform` class name should be `Vendor`. But, hum, I'm undecided.

> if ($session->vendorIs(Platform::MYSQL)) {

Author: OK, did it.

    // Write code for MySQL.
}
if ($session->vendorIs(Platform::POSTGRESQL)) {
    // Write code for PostgreSQL.
}
if ($session->vendorIs('postgresql')) {
    // Variant with a user given raw string.
}
if ($session->vendorIs('mysql')) {
    // Variant with a user given raw string.
}
```

And more importantly, all vendor name comparison methods accept a value
array for checking if the vendor is any of:

```php
use MakinaCorpus\QueryBuilder\DatabaseSession;
use MakinaCorpus\QueryBuilder\Platform;

\assert($session instanceof DatabaseSession);

if ($session->vendorIs([Platform::MARIADB, Platform::MYSQL)) {
    // Write code for MariaDB and MySQL.
}
```

:::info
It is advised to use the `MakinaCorpus\QueryBuilder\Platform` constants for
checking against the vendor name, yet is not mandatory: user given vendor
name is a raw string, which will be lowercased and reduced to only latin

> alphabet characters prior being compared against.

Reviewer: ```suggestion
alphabet characters prior to being compared against.
```

For example the `SQL Server 2022` value will be reduced to `sqlserver`.

Values will be compared against a set of known synonyms in order to be

> tolerant to user conventions, thought it is not advised to rely upon this
> tolerances since the known synonyms is a not public API and might change.

Reviewer: ```suggestion
tolerant to user conventions, though it is not advised to rely upon these
tolerances since the known synonyms are not part of the public API and
might change.
```
